docs/design-documents/20231117-better-processors.md

```diff
@@ -0,0 +1,889 @@
+	Specification() Specification
+
+	// Configure is the first function to be called in a processor. It provides the
+	// processor with the configuration that needs to be validated and stored.
+	// In case the configuration is not valid it should return an error.
+	// Configure should not open connections or any other resources. It should solely
+	// focus on parsing and validating the configuration itself.
+	Configure(context.Context, map[string]string) error
+
+	// Open is called after Configure to signal the processor it can prepare to
+	// start writing records. If needed, the processor should open connections and
+	// start background jobs in this function.
+	Open(context.Context) error
+
+	// Process takes a number of records and processes them right away.
+	// It should return a slice of ProcessedRecord that matches the length of
+	// the input slice. If an error occurred while processing a specific record
+	// it should be reflected in the ProcessedRecord with the same index as the
+	// input record that caused the error.
+	// Process should be idempotent, as it may be called multiple times with the
+	// same records (e.g. after a restart when records were not flushed).
+	Process(context.Context, []Record) ([]ProcessedRecord)
+
+	// Teardown signals to the processor that the pipeline is shutting down and
+	// there will be no more calls to any other function. After Teardown returns,
+	// the processor will be discarded.
+	Teardown(context.Context) error
+}
+```
+
+Notice that the processor returns its own specification in the first method. This
+is required so that Conduit can build a registry of all available processors
+and give the user more information about them (see
+[Listing processor plugins](#listing-processor-plugins)).
+
+A few notes about the `Process` method:
+
+- Even though we are not adding support for processing multiple records at the
+  same time, we propose to design the interface in a way that will let us add
+  that functionality in the future without breaking changes. This means that the
+  function will for now be called only with a single record.
+- Notice that the return value is a slice of type `ProcessedRecord`. The purpose
+  behind this is to be able to figure out what happened to each input record. This
+  way we can associate each input record to an output record and know what happened
+  to it (was it filtered out, split, merged, or only transformed). This distinction
+  is important because of acknowledgments (e.g. if a record is split into multiple
+  records, Conduit will acknowledge the original record only once all split records
+  are acknowledged).
+- If an error happens in `Process` it should be stored in the `ProcessedRecord` that
+  was being processed when the error occurred. In that case the processor has two
+  options on how to continue processing the remaining records in the slice.
+  1. If the processor has no side effects and cheap it can process the remaining
```

Comment:
are we the ones deciding if a processor is cheap or not?

Reply:
Yes, the more I think about it, we should just have option 2, return early and let Conduit decide what to do.